nomad/structs/volumes.go

```diff
@@ -0,0 +1,188 @@
+package structs
+
+import (
+	"github.com/mitchellh/copystructure"
+	"github.com/mitchellh/mapstructure"
+)
+
+const (
+	VolumeTypeHost = "host"
+)
+
+// ClientHostVolumeConfig is used to configure access to host paths on a Nomad Client
+type ClientHostVolumeConfig struct {
+	Name     string `hcl:",key"`
+	Source   string `hcl:"source"`
+	ReadOnly bool   `hcl:"read_only"`
+	Hidden   bool   `hcl:"hidden"`
+}
+
+func (p *ClientHostVolumeConfig) Copy() *ClientHostVolumeConfig {
+	if p == nil {
+		return nil
+	}
+
+	c := new(ClientHostVolumeConfig)
+	*c = *p
+	return c
+}
+
+func CopyMapStringClientHostVolumeConfig(m map[string]*ClientHostVolumeConfig) map[string]*ClientHostVolumeConfig {
+	if m == nil {
+		return nil
+	}
+
+	nm := make(map[string]*ClientHostVolumeConfig, len(m))
+	for k, v := range m {
+		nm[k] = v.Copy()
+	}
+
+	return nm
+}
+
+func HostVolumeSliceMerge(a, b []*ClientHostVolumeConfig) []*ClientHostVolumeConfig {
+	n := make([]*ClientHostVolumeConfig, len(a))
+	seenKeys := make(map[string]struct{}, len(a))
+
+	for k, v := range a {
+		if _, ok := seenKeys[v.Name]; ok {
+			continue
+		}
+		n[k] = v.Copy()
+		seenKeys[v.Name] = struct{}{}
+	}
+	for k, v := range b {
+		if _, ok := seenKeys[v.Name]; ok {
+			continue
```

Comment:
If both `a` and `b` contain a similarly named host volumes, what's rationale for dropping `b` version silently instead of `a`'s?  I would have expected last one to win.  Either way, should call that in the function documentation.